Feature: SDL Adapter

Berulacks opened the pull request:
SDL support should be ready. Needs testing on Windows to see if I broke anything on that platform.

dislogical committed:
Added basic SDL support, Dash now runs on OSX (but immediately crashes)

dislogical committed:
SDL adapter now creates an OpenGL context

dislogical committed:
Started conversion to OOP wrapper [skip ci]

dislogical committed:
Window now opens and renders! Also, you can quit, so that's good

dislogical committed:
Added dub.selections.json to make build faster

dislogical committed:
Changed dub versions to be specific

dislogical committed:
Fixed crashes on OSX

dislogical committed:
Final preparations for merge

* Removed Mac adapter, because it is no longer necessary.
* Setup Windows to use Win32, Linux to use Linux, and everything else to use SDL.

dislogical committed:
Added mouse click support, added UseSDL2 version to force usage

dislogical committed:
Deleted the selections file. More pain than it's worth

Berulacks committed:
Merge branch 'develop' of github.com:Circular-Studios/Dash into SDLAdapter

Conflicts:
	source/dash/graphics/adapters/package.d
	source/dash/graphics/graphics.d

Berulacks committed:
Improved SDL support to the point that it appears to be on par with
windows. Fixes #207

Berulacks committed:
Changed keys in state to an adaptive array

Berulacks committed:
Cleaned up unnecessary code. This fixes #207

Berulacks committed:
Merge branch 'develop' of https://github.com/Circular-Studios/Dash into SDLAdapter

Conflicts:
	.travis.yml
	dub.json
	source/dash/utility/output.d

dislogical commented on source/dash/graphics/shaders/shaders.d:
```diff
@@ -321,12 +321,17 @@ public:
-     {
+    final void bindUI( UserInterface ui )
+    {
+        // This is part of a bigger problem. But in the interest of dope screenshots...
```
I do love dope screenshots.

dislogical commented:
This appears to work on my Macbook, which is probably the best test I have for these platforms. LGTM.

brooklynlittell commented:
I'm ~~crashing~~ not building on Windows 8.1 (Macbook Pro). Here are my errors:

`..\Dash\source\dash\utility\input.d(219): Error: undefined identifier Win32, did you mean import win32?
..\Dash\source\dash\utility\input.d(220): Error: cannot implicitly convert expression (Vector(0)) of type Vector!(int, 2) to Vector!(uint, 2)`

dislogical commented on source/dash/utility/input.d:
```diff
@@ -215,6 +216,9 @@ public:
         version( Windows )
         {
+            if( !Win32.get() )
+                return vec2i();
```
This should be:

``` d
import dash.graphics.adapters.win32gl;
if( !Win32.get() )
    return vec2ui();
```

dislogical replied:
JK it should be:

``` d
if( !Win32GL.get() )
    return vec2ui();
```

dislogical commented:
@BrandonLittell can you try the fix I just commented to see if that fixes it?

brooklynlittell commented:
Yeah, I just ran it. Obviously the vec was fixed, but win32 is still undefined. :cry:

dislogical commented:
Did that fix it? You shouldn't need the `import`, just rename `Win32` to `Win32GL`.

brooklynlittell commented:
That fixed it, but we're now running out of page space while linking our .lib.

```
..\Dash\dash.lib
Warning 178: .LIB pagesize exceeds 512
```

Commit:
Fixed compiler errors.

dislogical commented:
That's new. I'm not sure that means we're out of space, it probably just has to start swapping stuff out of RAM to the disk. Which is maaaad slow.

Also, thanks for committing those fixes.

brooklynlittell commented:
It starts to copy files for vibe-d (`Copying files for vibe-d...`), then exits the command and returns to my shell.

dislogical commented:
Is that on a `dub run` or a `dub build`?

brooklynlittell commented:
Damn, it was a dub build. It runs now :wink:

Berulacks commented:
:D

dislogical commented:
Yeah, the `Copying files for xxx...` thing doesn't say when it finishes, it just ends.

@BrandonLittell if this works for you now, could you merge it?

brooklynlittell commented:
If I press the arrow keys the engine locks up, dies after maybe 30 seconds, and spits out `core.exception.OutOfMemoryError@(0)`. If you're all good with this feature, I'll merge asap.

dislogical commented:
@BrandonLittell Womp womp.
@Berulacks Maybe fix that?

Berulacks commented:
I'm on it.

Berulacks commented:
It'll have to be later in the day, when I get out of class and in front of a Windows pc. Does it only happen with the arrow keys?

brooklynlittell commented:
Command, alt, and control all seem to kill it, too.

Berulacks commented:
Ah, okay. That's what I was worried about. I'll fix it asap.

Berulacks committed:
Fixed opBinary in keyboard

(the type of 'keys' was changed to an associative array, rendering the
old for loop void and prone to breaking everything)

Berulacks committed:
Removed an unnecessary debug line

Berulacks commented:
Okay guys, should be working now! I finally got everything set up on my Windows machine, so I tested it as well. Under Windows (running through SDL2) the keys seem to be working fine.

#207 = fixed!

dislogical committed:
Fixed openssl version issues

dislogical committed:
Fixed spacing

dislogical commented:
Check your spacing, it looks like you were using hardware tabs set to 8 spaces?

Also, it seems like SDL is locking us in at 60FPS. Can we change this based on `config.graphics.vsync`?

Berulacks committed:
Fixed vsync in SDL

The SDL GL swap interval now properly reflects the value of
config.graphics.vsync

Berulacks committed:
Fixed spacing in input.d

dislogical commented:
Works fine for me on the lab machines. I'll give it a test on my linux box today.

@BrandonLittell Can you test this on your Macbook Pro (OSX)?
@PxlBuzzard Can you test this on your Windows 10 box?

Please report back with results.

PxlBuzzard commented:
Runs on Windows 10 Preview with no special configuration. One difference I noticed between develop and SDL:

Develop:
![dash-develop](https://cloud.githubusercontent.com/assets/512416/4581284/b38353e8-4fde-11e4-8c0d-9914dcf752ec.PNG)

SDL:
![dash-sdl](https://cloud.githubusercontent.com/assets/512416/4581285/b4eee670-4fde-11e4-8007-9878d50e225d.PNG)

I don't care one way or another about the buttons on the right, but it would be nice to get the icon back in SDL.

Berulacks commented:
Ah, I forgot about the icon. Let me get out of lunch and I'll fix that.

dislogical commented:
On my Linux laptop (2014 Razer Blade, Ubuntu 14.04), SDL.initialize() is failing on [line 58](https://github.com/Circular-Studios/Dash/blob/feature/SDLAdapter/source/dash/graphics/adapters/sdl.d#L58). Here's the stack trace:

```
derelict.util.exception.DerelictException@../../../../.dub/packages/derelict-util-1.0.3/source/derelict/util/exception.d(35): DerelictGL3.reload failure: An OpenGL context is not currently active.
----------------
./sample-dash-game(void dash.graphics.adapters.sdl.Sdl.initialize()+0x291) [0xa71d35]
./sample-dash-game(void dash.graphics.graphics.Graphics.initialize()+0x47) [0xa476f3]
./sample-dash-game(_D4dash4core5dgame5DGame5startMFZ9__lambda7FZv+0x9) [0xa7dc3d]
./sample-dash-game(core.time.TickDuration[1] std.datetime.benchmark!(_D4dash4core5dgame5DGame5startMFZ9__lambda7FZv).benchmark(uint)+0x6b) [0xa7dc1f]
./sample-dash-game(void dash.utility.output.bench!(_D4dash4core5dgame5DGame5startMFZ9__lambda7FZv).bench(lazy immutable(char)[])+0x1b) [0xa5b987]
./sample-dash-game(void dash.core.dgame.DGame.start()+0xe2) [0xa3fc52]
./sample-dash-game(void dash.core.dgame.DGame.run()+0x60) [0xa3f840]
./sample-dash-game(_Dmain+0x54) [0xa39554]
./sample-dash-game(_D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv+0x1f) [0xc541db]
./sample-dash-game(void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate())+0x2a) [0xc5412e]
./sample-dash-game(void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll()+0x30) [0xc54194]
./sample-dash-game(void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate())+0x2a) [0xc5412e]
./sample-dash-game(_d_run_main+0x193) [0xc540af]
./sample-dash-game(main+0x25) [0xa39b3d]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf5) [0x7fa57fef2ec5]
core.exception.InvalidMemoryOperationError@(0)
```

Berulacks commented:
@ColdenCullen What drivers are you using? Perhaps it's a derelictGL3 error? Also, mind if I save the icon as a .bmp as well? I'm assuming "LoadImage" is a win32 specific thing, so we'll have to use something like SDL_Image to load the icon, and SDL_Image doesn't support .icos (last I checked)

dislogical commented:
Oh yeah switching to the Nvidia drivers fixed it, whoops. Now I'm just getting a "FBX not supported" error, which sounds like a @tmr9209 problem. LGTM?

brooklynlittell commented:
Runs on my MacBook Pro with Windows 8.1. Special key presses don't kill the engine. :+1:

dislogical committed:
Added better version support to choosing an adapter

tmr9209 commented:
Sorry, have been busy with my Microsoft interview the past two days. Did you solve the issue or just work around it? On a quick glance that error message means that fbx has invalid data from assimp, although would have to be at PC to figure out the specific reason it errors now.

dislogical commented:
@tmr9209 It's definitely an issue with Assimp, the assert comes from [here](https://github.com/Circular-Studios/Dash/blob/feature/SDLAdapter/source/dash/components/assets.d#L78).

@BrandonLittell Any luck running on OSX?

@Berulacks Any updates on icon support? We're now 3 weeks behind on the release this was supposed to go out in.

Berulacks commented:
@ColdenCullen GFM [doesn't](https://github.com/d-gamedev-team/gfm/search?q=SDL_SetWindowIcon&type=Issues&utf8=%E2%9C%93) appear to implement [SDL_SetWindowIcon](https://wiki.libsdl.org/SDL_SetWindowIcon), and we can't access the member variables [_window](https://github.com/d-gamedev-team/gfm/blob/master/sdl2/gfm/sdl2/window.d#L461) and [_surface](https://github.com/d-gamedev-team/gfm/blob/master/sdl2/gfm/sdl2/surface.d#L170) in [SDL2Window](https://github.com/d-gamedev-team/gfm/blob/master/sdl2/gfm/sdl2/surface.d) and [SDL2Surface](https://github.com/d-gamedev-team/gfm/blob/master/sdl2/gfm/sdl2/surface.d) which we would need if we just wanted to call SDL_SetWindowIcon ourselves. I'm looking into it, maybe I'll send them a pull request...

[EDIT] Woops, we [can](https://github.com/d-gamedev-team/gfm/blob/master/sdl2/gfm/sdl2/surface.d#L128) get the _surface.

dislogical commented:
@Berulacks You have access to all of the C functions. Also: [handle](http://d-gamedev-team.github.io/gfm/gfm.sdl2.surface.html#SDL2Surface.handle)

Berulacks commented:
Yeah, I didn't notice handle until my second browse through.

Berulacks committed:
Added icon support for SDL

(searches for icon.bmp in Textures)
Also changed the version call in input.d to DashUseSDL2, as per the
change in graphics.d

Berulacks commented:
Done. [SDL_GetWindowFromID](https://wiki.libsdl.org/SDL_GetWindowFromID) saved the day! Though, the guys at gfm should probably add a SDL2Window.setIcon function.

dislogical commented:
PR it! I'm sure @p0nce will be happy to review it.

@PxlBuzzard Work for you?

PxlBuzzard commented:
Yes, the icon shows up now. :+1:

dislogical committed:
Updated gfm:sdl2 to use window.setIcon()
